File: ruby/lib/callisto/transport.rb
# frozen_string_literal: true

require "net/http"
require "uri"
require "json"

module Callisto
  # HTTP transport over the standard library Net::HTTP. Applies Basic auth, drops nil
  # query params and body fields, serializes JSON, and maps non-2xx responses to typed errors.
  class Transport
    # @param config [Config]
    def initialize(config)
      @config = config
    end

    # Performs an HTTP request and returns the decoded response body.
    #
    # @param method [String] HTTP verb ("GET", "POST", ...)
    # @param path [String] path appended to the base URL (e.g. "/sms/send")
    # @param body [Object, nil] request body, serialized as JSON when present
    # @param query [Hash, nil] query parameters; nil values are dropped
    # @return [Object, nil] decoded JSON (Hash/Array), raw String, or nil
    # @raise [CallistoError] on a non-2xx response or transport failure
    def request(method, path, body: nil, query: nil)
      uri = URI.parse(@config.base_url + path)
      params = (query || {}).reject { |_, v| v.nil? }
      uri.query = URI.encode_www_form(params) unless params.empty?

      req = build_request(method, uri, body)

      response = perform(uri, req)

      data = decode_body(response)

      if error_response?(response)
        status = response.code.to_i
        message =
          if data.is_a?(Hash) && (data.key?("message") || data.key?(:message))
            data["message"] || data[:message]
          else
            "HTTP #{status}"
          end
        retry_after = nil
        if status == 429
          raw = response["Retry-After"]
          retry_after = Integer(raw, 10) rescue nil unless raw.nil?
        end
        raise Callisto.error_from_status(status, message.to_s, data, retry_after)
      end

      data
    end

    # Net::HTTP holds no persistent connection in this design; provided for API symmetry.
    def close
      nil
    end

    private

    def build_request(method, uri, body)
      klass = case method.to_s.upcase
              when "GET" then Net::HTTP::Get
              when "POST" then Net::HTTP::Post
              when "PUT" then Net::HTTP::Put
              when "PATCH" then Net::HTTP::Patch
              when "DELETE" then Net::HTTP::Delete
              else raise ArgumentError, "Unsupported HTTP method: #{method}"
              end

      req = klass.new(uri)
      req["Accept"] = "application/json"
      credentials = "#{@config.client_id}:#{@config.api_key}"
      req["Authorization"] = "Basic " + [credentials].pack("m0")

      unless body.nil?
        req["Content-Type"] = "application/json"
        req.body = JSON.generate(body)
      end

      req
    end

    def perform(uri, req)
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = uri.scheme == "https"
      http.open_timeout = @config.timeout
      http.read_timeout = @config.timeout
      http.write_timeout = @config.timeout if http.respond_to?(:write_timeout=)
      http.request(req)
    rescue StandardError => e
      raise NetworkError.new("Request to #{uri} failed: #{e.message}")
    end

    def decode_body(response)
      raw = response.body
      return nil if raw.nil? || raw.empty?

      begin
        JSON.parse(raw)
      rescue JSON::ParserError
        raw
      end
    end

    def error_response?(response)
      response.code.to_i >= 400
    end
  end
end

File: java/src/main/java/com/callisto/sdk/http/Transport.java
package com.callisto.sdk.http;

import com.callisto.sdk.Config;
import com.callisto.sdk.errors.CallistoException;
import com.callisto.sdk.errors.Errors;
import com.callisto.sdk.errors.NetworkException;
import com.callisto.sdk.reporting.ErrorReporter;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * HTTP transport layer.
 *
 * <p>Applies HTTP Basic auth ({@code Authorization: Basic base64(clientId:apiKey)}),
 * an {@code Accept: application/json} header, serializes request bodies as JSON,
 * drops {@code null} query params, parses error bodies, and maps status codes to
 * typed exceptions.
 */
public class Transport {

    private final Config config;
    private final HttpClient httpClient;
    private final ObjectMapper mapper;
    private final String authHeader;
    private ErrorReporter reporter;

    public Transport(Config config) {
        this(config, defaultClient(config), new ObjectMapper());
    }

    /**
     * Constructs a transport with an injectable {@link HttpClient} (for testing) and
     * {@link ObjectMapper}.
     */
    public Transport(Config config, HttpClient httpClient, ObjectMapper mapper) {
        this.config = config;
        this.httpClient = httpClient;
        this.mapper = mapper;
        String credentials = config.getClientId() + ":" + config.getApiKey();
        this.authHeader = "Basic " + Base64.getEncoder()
                .encodeToString(credentials.getBytes(StandardCharsets.UTF_8));
    }

    /**
     * Attaches the shared error reporter so the transport (the single API/network error choke
     * point) and resources can capture before throwing. Optional; {@code null} disables hooks.
     */
    public void setReporter(ErrorReporter reporter) {
        this.reporter = reporter;
    }

    /** The shared error reporter, or {@code null} when reporting is disabled. */
    public ErrorReporter reporter() {
        return reporter;
    }

    /**
     * Captures an exception via the reporter (if any), tagging it with the originating HTTP
     * {@code method} and {@code path} so the reporter sets {@code culprit} and {@code request}.
     * Never throws.
     */
    public void capture(Throwable t, String method, String path) {
        if (reporter == null) {
            return;
        }
        Map<String, Object> extra = new LinkedHashMap<>();
        if (method != null) {
            extra.put("__method", method);
        }
        if (path != null) {
            extra.put("__path", path);
        }
        reporter.captureException(t, "error", extra);
    }

    private static HttpClient defaultClient(Config config) {
        return HttpClient.newBuilder()
                .connectTimeout(config.getTimeout())
                .build();
    }

    public ObjectMapper mapper() {
        return mapper;
    }

    /**
     * Executes a request and returns the decoded JSON response body as a {@link JsonNode}
     * (or {@code null} for an empty body).
     *
     * @param method HTTP method.
     * @param path   request path appended to the base URL.
     * @param body   request body (serialized as JSON) or {@code null}.
     * @param query  query parameters; entries with {@code null} values are dropped.
     */
    public JsonNode request(String method, String path, Object body, Map<String, Object> query) {
        String url = config.getBaseUrl() + path + buildQueryString(query);

        HttpRequest.Builder builder = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .timeout(config.getTimeout())
                .header("Authorization", authHeader)
                .header("Accept", "application/json");

        HttpRequest.BodyPublisher publisher;
        if (body != null) {
            String json;
            try {
                json = mapper.writeValueAsString(body);
            } catch (IOException e) {
                throw new NetworkException("Failed to serialize request body: " + e.getMessage(), e);
            }
            publisher = HttpRequest.BodyPublishers.ofString(json, StandardCharsets.UTF_8);
            builder.header("Content-Type", "application/json");
        } else {
            publisher = HttpRequest.BodyPublishers.noBody();
        }
        builder.method(method, publisher);

        HttpResponse<String> response;
        try {
            response = httpClient.send(builder.build(), HttpResponse.BodyHandlers.ofString());
        } catch (IOException e) {
            NetworkException ex = new NetworkException("Request to " + url + " failed: " + e.getMessage(), e);
            capture(ex, method, path);
            throw ex;
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            NetworkException ex =
                    new NetworkException("Request to " + url + " was interrupted: " + e.getMessage(), e);
            capture(ex, method, path);
            throw ex;
        }

        int status = response.statusCode();
        String rawBody = response.body();

        JsonNode data = null;
        if (rawBody != null && !rawBody.isEmpty()) {
            try {
                data = mapper.readTree(rawBody);
            } catch (IOException e) {
                // Non-JSON body: surface it as a text node.
                data = mapper.getNodeFactory().textNode(rawBody);
            }
        }

        if (status < 200 || status >= 300) {
            String message;
            if (data != null && data.isObject() && data.has("message")) {
                message = data.get("message").asText();
            } else {
                message = "HTTP " + status;
            }
            Integer retryAfter = null;
            if (status == 429) {
                retryAfter = response.headers().firstValue("Retry-After")
                        .map(raw -> {
                            try {
                                return Integer.parseInt(raw.trim());
                            } catch (NumberFormatException ex) {
                                return null;
                            }
                        })
                        .orElse(null);
            }
            CallistoException ex = Errors.fromStatus(status, message, data, retryAfter);
            capture(ex, method, path);
            throw ex;
        }
        return data;
    }

    private String buildQueryString(Map<String, Object> query) {
        if (query == null || query.isEmpty()) {
            return "";
        }
        Map<String, Object> filtered = new LinkedHashMap<>();
        for (Map.Entry<String, Object> e : query.entrySet()) {
            if (e.getValue() != null) {
                filtered.put(e.getKey(), e.getValue());
            }
        }
        if (filtered.isEmpty()) {
            return "";
        }
        StringBuilder sb = new StringBuilder("?");
        boolean first = true;
        for (Map.Entry<String, Object> e : filtered.entrySet()) {
            if (!first) {
                sb.append("&");
            }
            first = false;
            sb.append(encode(e.getKey()))
                    .append("=")
                    .append(encode(String.valueOf(e.getValue())));
        }
        return sb.toString();
    }

    private static String encode(String s) {
        return URLEncoder.encode(s, StandardCharsets.UTF_8);
    }

    public void close() {
        // java.net.http.HttpClient has no explicit close on Java 11; nothing to release.
    }
}
